### ISSUE TITLE
VPN fails in loop with wrong credentials when DNS failure happens

### ISSUE THREAD

**Issue by henriquevcosta** [binhex/arch-qbittorrentvpn#242]
I have recently deployed this image one a new system and have seen that it's having trouble with OpenVPN authentication after some DNS failures.
In the code log below I've REDACTED the IP addresses. After this, the last portion about the authentication part repeats multiple times a second and it just doesn't progress or recover (and generates a couple GB of logs in about 3 hours!).
I guess I can hardcode the VPN credentials in a file and change my .ovpn to point to it, but I take the problem to be an indication of some condition where the credentials are not being correctly provided to the openvpn command.

This is now the second time I've found the container in this state and can fix it by restarting. 

The DNS failure itself is not related to this, it happened externally to the container.

```
2024-06-05 08:31:07 VERIFY EKU OK
2024-06-05 08:31:07 VERIFY OK: depth=0, C=SE, ST=CA, (...)

2024-06-05 08:31:10,000 DEBG 'start-script' stdout output:
2024-06-05 08:31:10 Control Channel: TLSv1.3, cipher TLSv1.3 TLS_AES_256_GCM_SHA384, peer certificate: 2048 bits RSA, signature: RSA-SHA256, peer temporary key: 253 bits X25519

2024-06-05 09:16:41,465 DEBG 'watchdog-script' stdout output:
[info] DNS failure, creating file '/tmp/dnsfailure' to indicate failure...

2024-06-05 09:17:08,307 DEBG 'start-script' stdout output:
[info] Sending SIGTERM (-15) to 'openvpn' due to dns failure...

2024-06-05 09:17:08,336 DEBG 'start-script' stdout output:
2024-06-05 09:17:08 event_wait : Interrupted system call (fd=-1,code=4)

2024-06-05 09:17:08,337 DEBG 'start-script' stdout output:
2024-06-05 09:17:08 SIGTERM received, sending exit notification to peer

2024-06-05 09:17:10,021 DEBG 'start-script' stdout output:
2024-06-05 09:17:10 net_route_v4_del: REDACTED dev [NULL] table 0 metric -1
2024-06-05 09:17:10 net_route_v4_del: REDACTED dev [NULL] table 0 metric -1
2024-06-05 09:17:10 net_route_v4_del: REDACTED dev [NULL] table 0 metric -1

2024-06-05 09:17:10,021 DEBG 'start-script' stdout output:
2024-06-05 09:17:10 Closing TUN/TAP interface
2024-06-05 09:17:10 net_addr_v4_del: REDACTED dev tun0

2024-06-05 09:17:10,037 DEBG 'start-script' stdout output:
2024-06-05 09:17:10 SIGTERM[soft,exit-with-notification] received, process exiting

2024-06-05 09:17:10,069 DEBG 'start-script' stdout output:
[info] Starting OpenVPN (non daemonised)...

2024-06-05 09:17:10,090 DEBG 'start-script' stdout output:
2024-06-05 09:17:10 Note: option tun-ipv6 is ignored because modern operating systems do not need special I
Pv6 tun handling anymore.
2024-06-05 09:17:10 WARNING: Compression for receiving enabled. Compression has been used in the past to br
eak encryption. Sent packets are not compressed unless "allow-compression yes" is also set.

2024-06-05 09:17:10,091 DEBG 'start-script' stdout output:
2024-06-05 09:17:10 Note: '--allow-compression' is not set to 'no', disabling data channel offload.

2024-06-05 09:17:10,091 DEBG 'start-script' stdout output:
2024-06-05 09:17:10 OpenVPN 2.6.8 [git:makepkg/3b0d9489cc423da3+] x86_64-pc-linux-gnu [SSL (OpenSSL)] [LZO]
 [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] [DCO] built on Nov 17 2023
2024-06-05 09:17:10 library versions: OpenSSL 3.2.0 23 Nov 2023, LZO 2.10
2024-06-05 09:17:10 DCO version: N/A

2024-06-05 09:17:10,092 DEBG 'start-script' stdout output:
2024-06-05 09:17:10 neither stdin nor stderr are a tty device and you have neither a controlling tty nor sy
stemd - can't ask for 'Enter Auth Username:'.  If you used --daemon, you need to use --askpass to make pass
phrase-protected keys work, and you can not use --auth-nocache.
```

**Comment by binhex** [comment]
The snippet above is not showing wrong credentials, what it is telling you is that your openvpn file does not include `auth-user-pass <filename>` and thus openvpn is prompting to enter the username and password, but because there is no stdin as this is a container it cannot continue.

**Comment by henriquevcosta** [comment]
Thanks for the quick reply! 
I get that, which is why I was saying that providing the credentials to openvpn directly could be an option, but I was assuming that if the credentials provided with the env vars `VPN_USER` and `VPN_PASS` work with the container up until the moment of the DNS failure some hours later, they should keep working after DNS is back up. 
Or are those env vars not meant to work at all and I've just been lucky so far?
 My setup (this is Ansible) is:

```
    env:
      PGID: "{{ plex_groupid | quote }}"
      PUID: "{{ plex_userid | quote }}"
      VPN_ENABLED: "yes"
      VPN_USER: "{{ VPN_USERNAME }}"
      VPN_PASS: "{{ VPN_PASSWORD }}"
      VPN_PROV: "custom"
      VPN_CLIENT: "openvpn"
      STRICT_PORT_FORWARD: "no"
      ENABLE_PRIVOXY: "yes"
      ENABLE_AUTODL_IRSSI: "no"
      DEBUG: "false"
```

and then I have an opvn file (no credentials) in the `.../config/openvpn` directory.

**Comment by binhex** [comment]
> if the credentials provided with the env vars VPN_USER and VPN_PASS work with the container up until the moment of the DNS failure some hours later, they should keep working after DNS is back up.

It's possible that the DNS failure means openvpn is unable to authenticate with the endpoint, but what is a little confusing is why it would then prompt for the credentials, in any case can you please check the below;-

1. Missing `auth-user-pass` in the openvpn configuration file - This should not be the case as i handle this for you and force the insertion of this line into your openvpn config file (please check its present).
2. Password protected certificate used for authentication - Fix is to remove the password protection from the cert

**Comment by binhex** [comment]
Another question to ask is why are you getting name server lookup failure, what is the value of env var `NAME_SERVERS`

**Comment by henriquevcosta** [comment]
Oh I wasn't aware that you modified the file after startup - I suspect that may be the issue. If the container was already running my deployment scripts will put the file in place but not restart the container, so if the openvpn process crashes the file will not have the modifications anymore.

May I suggest then a small change which is that you don't modify the openvpn file in place, but instead create a modified copy so that if the user modifies the file that's mapped into the container it isn't picked up? 
Or maybe just add a note to the documentation...

Anyway at the moment I have the container running and:

1. There was a dns failure and no issue - corroborates my intepretation above, since I haven't reexecuted my scripts since the container started
2. I'm not sure how to check - I see the openvpn key in the ovpn file but don't know how to check if it has a password
3. I'm setting NAME_SERVERS to `"127.0.0.11,1.1.1.1"` so that I can use docker's DNS to have qbittorrent resolve my jackett container's hostname to allow for searching from inside qbittorrent

**Comment by binhex** [comment]
>  I suspect that may be the issue. If the container was already running my deployment scripts will put the file in place but not restart the container

Not really sure why you are doing this, can you not just simply copy the ovpn file to /config/openvpn/ and then make any changes manually that you want to make, that way it can then be edited correctly by my startup script, no need to keep copying the ovpn file.

> I'm setting NAME_SERVERS to "127.0.0.11,1.1.1.1" so that I can use docker's DNS to have qbittorrent resolve my jackett container's hostname to allow for searching from inside qbittorrent

OK that also explains the DNS failure message, 127.0.0.1 will not be able to resolve external hostnames, and no, DNS will not automatically move onto using 1.1.1.1 that's not how it works, DNS will only flip to the next name server in the list if the name server is down, if the name server is accessible but not able to resolve (in your case) then you will get a name server lookup failure.

Instead an alternative way to do this is to set qbittorrent to talk to jackett via your hosts static ip (it is static right?, it should be), that is how i have it setup, as long as your hosts ip doesn't change you are good, sure a hostname is the ideal but this works and in my experience is robust.

**Comment by henriquevcosta** [comment]
The scripts that place the file there are part of an ansible playbook that connect to my server and set up multiple things. They try as best as possible to be idempotent and they represent the intended state - hence they were modifying the ovpn file on every deployment. I can adjust the scripts to not copy every time, no worries. But maybe it could be highlighted in the docs that the file will be modified?

Regarding DNS I'm confused, the docker DNS server should forward queries to whatever's set in the host's /etc/resolv.conf as [per the docs](https://docs.docker.com/network/#dns-services) so there must be something else broken either in the host OS or in my docker engine config. I knew what you said about the multiple DNS servers, 1.1.1.1 was there only as a fallback in case something was wrong with docker's.

I'll keep investigating the DNS part but for the purpose of this issue I think the main of it was the file editing - now that I know I was breaking the file by modifying it without restarting the container I can work around it.
Thank you for all your help.

One final question, though: if I put in /config/openvpn an `ovpn` file already with the `auth-user-pass` directive and another file with the credentials, will you still modify the ovpn file?

**Comment by binhex** [comment]
> But maybe it could be highlighted in the docs that the file will be modified?

Sure

> Regarding DNS I'm confused, the docker DNS server should forward queries to whatever's set in the host's /etc/resolv.conf 

So I take control of name resolution to prevent any possibility of IP leakage via name server query, the script deletes all entries in /etc/resolv.conf and replaces it with whatever is specified by `NAME_SERVERS`, this then ensures no name resolution via local name server (iptables also blocks all name resolution on the LAN).

> I'll keep investigating the DNS part but for the purpose of this issue I think the main of it was the file editing

This will need to be fixed, i have scripts in place that check name resolution is operational, if it isn't then the container will continually restart the openvpn process.

>  if I put in /config/openvpn an ovpn file already with the auth-user-pass directive and another file with the credentials, will you still modify the ovpn file?

To keep things simple do not put in auth-user-pass directive, just ignore that and my script will fix it up.

**Comment by henriquevcosta** [comment]
Thanks for the reply, my edit makes sense then - it's probably the iptables blocking the outward call towards my router. 

So, taking a step back, and sorry for the ignorance, but with all things being up and running should the DNS queries be going out via the tunnel interface to avoid leakage? 
If that's the idea then it would indeed never make sense to use the docker DNS server, which takes me back to your suggestion of referring to jackett by IP instead of hostname...

**Comment by binhex** [comment]
>  but with all things being up and running should the DNS queries be going out via the tunnel interface to avoid leakage?

Yes exactly, as mentioned above i only permit name server lookup for the endpoints over the LAN, i then store this information in the hosts file inside of the container and from then onwards name server lookup is only permitted over the vpn tunnel, it is actively blocked on the LAN.

**Comment by henriquevcosta** [comment]
Thanks for the help, I'll move to fixed IPs and use the default DNS servers. I think we can close the issue. 
I'm having some problems with privoxy but I'll open a separate issue.
